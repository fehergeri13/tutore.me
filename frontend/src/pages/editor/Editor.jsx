import React from 'react';
import {observable, toJS} from 'mobx';
import {observer, inject} from 'mobx-react';
import autobind from 'autobind-decorator';
import "./editor.less";

@inject('model')
@observer
export default class Editor extends React.Component {

    async componentDidMount() {
        if (this.props.id !== undefined) {
            this.isLoading = true;

            const post = await this.props.model.rest.getPost(this.props.id);
            const keys = Object.keys(toJS(this.postData));

            keys.forEach(key => {
                this.postData[key] = post[key];
            });

            this.isLoading = false;
        }
    }

    @observable isLoading = false;

    @observable postData = {
        name: "",
        body: "",
        type: "",
        subject: [],
    };

    @autobind
    handleFormChange(e) {
        const target = e.target;
        const value = target.type === 'checkbox' ? target.checked : target.value;
        const name = target.name;

        this.postData[name] = value;

        console.log('postData', toJS(this.postData))
    }

    @autobind
    handleMultiCheckboxChange(e) {
        const target = e.target;
        const checked = target.checked;
        const value = target.value;
        const name = target.name;

        if (checked) {
            this.postData[name].push(value);
        } else {
            this.postData[name].remove(value);
        }

        console.log('postData', toJS(this.postData));
    }

    @autobind
    async handleFormSubmit(e) {
        e.preventDefault();

        try {
            if (this.props.id === undefined) {
                await this.props.model.rest.createPost(toJS(this.postData));
            } else {
                const data = {
                    id: this.props.id,
                    ...toJS(this.postData),
                };

                await this.props.model.rest.updatePost(data);
            }
        } catch (e) {
            console.error(e);
            alert(JSON.stringify(e))
        }

        this.props.model.routingStore.push('/user');
    }

    @autobind
    handleClearClick(e) {
        e.preventDefault();

        this.postData = {
            name: "",
            body: "",
            type: "",
            subject: [],
        };
    }

    renderForm() {
        if(this.isLoading) {
            return <h3>Adatok betöltése...</h3>
        }

        return <form onSubmit={this.handleFormSubmit}>
            <label className="title">
                <span>Hirdetés neve</span>
                <input
                    type="text"
                    name="name"
                    onChange={this.handleFormChange}
                    value={this.postData.name}
                />
            </label>
            <label className="text"><span>Hirdetés szövege</span>
                <textarea
                    name="body"
                    onChange={this.handleFormChange}
                    value={this.postData.body}
                />
            </label>

            <fieldset className="type">
                <legend>Típus</legend>
                <label>
                    <input
                        type="radio"
                        name="type"
                        value="demand"
                        checked={this.postData.type === 'demand'}
                        onChange={this.handleFormChange}
                    />
                    <span>Keres</span>
                </label>
                <label>
                    <input
                        type="radio"
                        name="type"
                        value="supply"
                        checked={this.postData.type === 'supply'}
                        onChange={this.handleFormChange}
                    />
                    <span>Kínál</span>
                </label>
            </fieldset>

            <fieldset className="subject">
                <legend>Tantárgy</legend>
                <label>
                    <input
                        type="checkbox"
                        name="subject"
                        value="informatika"
                        onChange={this.handleMultiCheckboxChange}
                        checked={this.postData.subject.includes('informatika')}
                    />
                    <span>Informatika</span>
                </label>
                <label>
                    <input
                        type="checkbox"
                        name="subject"
                        value="matematika"
                        onChange={this.handleMultiCheckboxChange}
                        checked={this.postData.subject.includes('matematika')}
                    />
                    <span>Matematika</span>
                </label>
                <label>
                    <input
                        type="checkbox"
                        name="subject"
                        value="fizika"
                        onChange={this.handleMultiCheckboxChange}
                        checked={this.postData.subject.includes('fizika')}
                    />
                    <span>Fizika</span>
                </label>
                <label>
                    <input
                        type="checkbox"
                        name="subject"
                        value="kemia"
                        onChange={this.handleMultiCheckboxChange}
                        checked={this.postData.subject.includes('kemia')}
                    />
                    <span>Kémia</span>
                </label>
            </fieldset>

            <div className="buttons">
                <button type="button" onClick={this.handleClearClick}>Törlés</button>
                <button type="submit">Mentés</button>
            </div>
        </form>;
    }

    render() {
        return <div className="editor-page">
            <div className="editor">
                <h2>Új hirdetés felvétele</h2>


                {this.renderForm()}
            </div>

        </div>;
    }
}
