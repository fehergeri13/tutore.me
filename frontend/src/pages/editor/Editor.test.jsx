import React from 'react';
import renderer from 'react-test-renderer';
import Editor from "./Editor";
import {Provider} from "mobx-react";

describe('Editor', () => {
    it('renders correctly', () => {
        const tree = renderer.create(
            <Provider model={{}}><Editor /></Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
