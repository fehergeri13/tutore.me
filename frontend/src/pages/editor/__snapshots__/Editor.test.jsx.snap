// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor renders correctly 1`] = `
<div
  className="editor-page"
>
  <div
    className="editor"
  >
    <h2>
      Új hirdetés felvétele
    </h2>
    <form
      onSubmit={[Function]}
    >
      <label
        className="title"
      >
        <span>
          Hirdetés neve
        </span>
        <input
          name="name"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <label
        className="text"
      >
        <span>
          Hirdetés szövege
        </span>
        <textarea
          name="body"
          onChange={[Function]}
          value=""
        />
      </label>
      <fieldset
        className="type"
      >
        <legend>
          Típus
        </legend>
        <label>
          <input
            checked={false}
            name="type"
            onChange={[Function]}
            type="radio"
            value="demand"
          />
          <span>
            Keres
          </span>
        </label>
        <label>
          <input
            checked={false}
            name="type"
            onChange={[Function]}
            type="radio"
            value="supply"
          />
          <span>
            Kínál
          </span>
        </label>
      </fieldset>
      <fieldset
        className="subject"
      >
        <legend>
          Tantárgy
        </legend>
        <label>
          <input
            checked={false}
            name="subject"
            onChange={[Function]}
            type="checkbox"
            value="informatika"
          />
          <span>
            Informatika
          </span>
        </label>
        <label>
          <input
            checked={false}
            name="subject"
            onChange={[Function]}
            type="checkbox"
            value="matematika"
          />
          <span>
            Matematika
          </span>
        </label>
        <label>
          <input
            checked={false}
            name="subject"
            onChange={[Function]}
            type="checkbox"
            value="fizika"
          />
          <span>
            Fizika
          </span>
        </label>
        <label>
          <input
            checked={false}
            name="subject"
            onChange={[Function]}
            type="checkbox"
            value="kemia"
          />
          <span>
            Kémia
          </span>
        </label>
      </fieldset>
      <div
        className="buttons"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          Törlés
        </button>
        <button
          type="submit"
        >
          Mentés
        </button>
      </div>
    </form>
  </div>
</div>
`;
